ddht/v5_1/alexandria/network.py

```diff
@@ -248,6 +314,81 @@ async def _serve_find_nodes(self) -> None:
                     request_id=request.request_id,
                 )
 
+    async def _ping_oldest_routing_table_entry(self) -> None:
+        await self._routing_table_ready.wait()
+
+        while self.manager.is_running:
+            # Here we preserve the lazy iteration while still checking that the
+            # iterable is not empty before passing it into `min` below which
+            # throws an ambiguous `ValueError` otherwise if the iterable is
+            # empty.
+            nodes_iter = self.routing_table.iter_all_random()
+            try:
+                first_node_id = first(nodes_iter)
+            except StopIteration:
+                await trio.sleep(ROUTING_TABLE_KEEP_ALIVE)
+                continue
+            else:
+                least_recently_ponged_node_id = min(
+                    cons(first_node_id, nodes_iter),
+                    key=lambda node_id: self._last_pong_at.get(node_id, 0),
+                )
+
+            too_old_at = trio.current_time() - ROUTING_TABLE_KEEP_ALIVE
+            try:
+                last_pong_at = self._last_pong_at[least_recently_ponged_node_id]
+            except KeyError:
+                pass
+            else:
+                if last_pong_at > too_old_at:
+                    await trio.sleep(last_pong_at - too_old_at)
+                    continue
+
+            did_bond = await self.bond(least_recently_ponged_node_id)
+            if not did_bond:
+                self.routing_table.remove(least_recently_ponged_node_id)
+
+    async def _track_last_pong(self) -> None:
+        async with self.client.subscribe(PongMessage) as subscription:
+            async for message in subscription:
+                self._last_pong_at[message.sender_node_id] = trio.current_time()
+
+    async def _manage_routing_table(self) -> None:
+        # First load all the bootnode ENRs into our database
+        for enr in self._bootnodes:
+            try:
+                self.enr_db.set_enr(enr)
+            except OldSequenceNumber:
+                pass
+
+        # Now repeatedly try to bond with each bootnode until one succeeds.
+        async with trio.open_nursery() as nursery:
+            while self.manager.is_running:
+                for enr in self._bootnodes:
+                    if enr.node_id == self.local_node_id:
+                        continue
+                    endpoint = Endpoint.from_enr(enr)
+                    nursery.start_soon(self._bond, enr.node_id, endpoint)
+
+                with trio.move_on_after(10):
+                    await self._routing_table_ready.wait()
+                    break
+
+        # TODO: Need better logic here for more quickly populating the
+        # routing table.  Should start off aggressively filling in the
+        # table, only backing off once the table contains some minimum
+        # number of records **or** searching for new records fails to find
+        # new nodes.  Maybe use a TokenBucket
```

Comment:
By doing a couple random lookups we *should* be able to pretty quickly estimate of how large the network is, using similar logic to how a loglog counter works. From there we can know around how many nodes we expect a "full" routing table to have in each bucket, and start backing off once our distribution reaches that expected distribution.